### PR TITLE
allow setting patches file location through `--patches-file` option (fixes #25)

### LINKED ISSUE #25: [feature] allow using dedicated `patches.json` file instead of nesting in `composer.json`
As outlined in `cweagans/composer-patches` [documentation](https://docs.cweagans.net/composer-patches/usage/defining-patches/#locations), the patch list can either be included directly in `composer.json` or in a dedicated `patches.json` file - or even [any JSON file as long as it is specified](https://docs.cweagans.net/composer-patches/usage/configuration/#patches-file) in `composer.json`'s `extra.patches-file` property.

A dedicated file is useful to remove some bloat from `composer.json`, however, this package seem to always place new patches directly in `composer.json`. It would be interested if it was possible to override this behavior and even adopt some smart assumptions. So I would propose a few changes to allow that:

1. Change default behavior when deciding where to add the patch entry:
	- if `extra.patches-file` is set in `composer.json`, use that field's value
	- else, if `patches.json` exists, use that
	- else, fallback to `composer.json`
2. Add a `--patches-file` option that allows overriding default behavior

Default behavior suggestion is based on a) following `cweagans/composer-patches` configuration and b) assuming if `patches.json` exist, it must be being used, but I'm open 'to hearing your thoughts on this.

I'd be happy to work on a PR to implement this change, just let me know what you think.

### PR DESCRIPTION
**Pull request by rmobis**
Adds `--patches-file` option which allows using a [patches file](https://docs.cweagans.net/composer-patches/usage/defining-patches/#locations) instead of relying on `composer.json`.

Since the structure for the `patches.json` is "shallower" (no `extra` at the top) the code path is a bit different. Could have done a slightly bigger refactor, but I like to try and keep things simple when contributing to someone else's code.

When `--patches-file` is set we assume this shallower format, meaning you can't call with `--patches-file=composer.json`. Could instead check the file name at runtime and decide what format to use, but this was simpler and I think it makes sense. Let me know what you think.

Also add a test for changes.

### CODE REVIEW

**Comment by rmobis** [comment]
I'm not really sure what the issue with rector is, it wasn't working for me locally even before I made any changes either. If that's something I did wrong, please let me know.

**Comment by samsonasik** [comment]
Rector in composer.json still on 0.18.3

https://github.com/symplify/vendor-patches/blob/585e7504e9f32f03186c2a39a3d530a95851616c/composer.json#L20

Could you try upgrade  to ^1.2.5 ? Thank you.

File: src/Composer/ComposerPatchesConfigurationUpdater.php [hunk 1]
```diff
@@ -56,4 +71,15 @@ public function updateComposerJsonAndPrint(string $composerJsonFilePath, array $
...
+
+    /**
+     * @param mixed[] $composerExtraPatches
+     */
+    public function updatePatchesFileJsonAndPrint(string $patchesFilePath, array $composerExtraPatches): void
+    {
+        $patchesFileJson = $this->updatePatchesFileJson($patchesFilePath, $composerExtraPatches);
+
+        $patchesFileContents = Json::encode($patchesFileJson, Json::PRETTY);
+        FileSystem::write($patchesFilePath, $patchesFileContents);
```
Comment (samsonasik) [inline src/Composer/ComposerPatchesConfigurationUpdater.php:83]: pass null on 3rg argument to keep permission

```suggestion
        FileSystem::write($patchesFilePath, $patchesFileContents, null);
```

**Comment by rmobis** [comment]
Please don't merge, I just noticed there's a bug. I'll fix it in a couple of hours when I back.

**Comment by rmobis** [comment]
Fixes have been applied, sorry for the delay. Let me know if there's anything else missing.

File: src/Command/GenerateCommand.php [hunk 2]
```diff
@@ -18,6 +19,8 @@
 
 final class GenerateCommand extends Command
 {
+    public const PATCHES_FILE_OPTION = 'patches-file';
```
Comment (samsonasik) [inline src/Command/GenerateCommand.php:22]: can this be private const?

Comment (rmobis) [reply src/Command/GenerateCommand.php]: Done

**Review by samsonasik — APPROVED** [review]

**Comment by rmobis** [comment]
@TomasVotruba is anything missing for this to be merged?

**Comment by TomasVotruba** [comment]
All good. Could you update the readme as well?

**Comment by TomasVotruba** [comment]
Thank you 😊

**Pull request merged**